runtime/src/main/scala/com/soundcloud/twinagle/ProtoService.scala

```diff
@@ -1,24 +1,32 @@
 package com.soundcloud.twinagle
 
-import com.twitter.finagle.Service
+import com.twitter.finagle.{Filter, Service}
 import com.twitter.finagle.http.{Request, Response}
 import com.twitter.util.Future
 import scalapb.{GeneratedMessage, GeneratedMessageCompanion}
 
-case class ProtoService(rpcs: Seq[ProtoRpc]) {
-  assert(rpcs.map(_.metadata.service).toSet.size == 1, "inconsistent services in metadata")
-}
+case class ProtoService(rpcs: Seq[ProtoRpcBuilder])
+
 object ProtoService {
   implicit val asProtoService: AsProtoService[ProtoService] = (t: ProtoService) => t
 }
 
 case class ProtoRpc(metadata: EndpointMetadata, svc: Service[Request, Response])
-object ProtoRpc {
+
+trait ProtoRpcBuilder {
+  def build(generatedMessageFilters: Seq[MessageFilter]): ProtoRpc
+}
+
+object ProtoRpcBuilder {
   def apply[
       Req <: GeneratedMessage: GeneratedMessageCompanion,
       Resp <: GeneratedMessage: GeneratedMessageCompanion
-  ](endpointMetadata: EndpointMetadata, rpc: Req => Future[Resp]): ProtoRpc = {
-    val httpService = new TwirpEndpointFilter[Req, Resp] andThen Service.mk(rpc)
-    ProtoRpc(endpointMetadata, httpService)
-  }
+  ](endpointMetadata: EndpointMetadata, rpc: Req => Future[Resp]): ProtoRpcBuilder =
+    (generatedMessageFilters: Seq[MessageFilter]) => {
+      val twirpFilter: Filter[Request, Response, Req, Resp] = new TwirpEndpointFilter[Req, Resp]
+      val svc = generatedMessageFilters.foldLeft(twirpFilter) { case (accFilter, nextFilter) =>
+        accFilter.andThen(nextFilter.toFilter)
+      } andThen Service.mk(rpc)
```

Comment:
I am not sure that I am following, I think we need to insert message filters between `TwirpEndpointFilter` and `Service.mk(rpc)`?

Reply:
sorry for the super-late reply! you're right about the positions of the filters. What I was aiming for was building up a `Service[Req, Resp]` using `foldRight` instead of building the `Filter[Request, Response, Req, Resp]`:

```scala
      val svc: Service[Req, Resp] = messageFilters.foldRight(Service.mk(rpc)) { case (messageFilter, accSvc) =>
        messageFilter.toFilter[Req, Resp] andThen accSvc
      }
      ProtoRpc(endpointMetadata, new TwirpEndpointFilter[Req, Resp] andThen svc)
```

That said, if we only keep a single `MessageFilter` in the builder (and maybe allow combining `MessageFilter`s via `andThen` in the future), it becomes a moot point :)